File: html_stripper.py
# html_stripper.py
import re
import html
import sys
from typing import Tuple, Optional

class HTMLStripper:
    def __init__(self, strip_comments=True, convert_br=True, decode_entities=True, collapse_spaces=True):
        self.strip_comments = strip_comments
        self.convert_br = convert_br
        self.decode_entities = decode_entities
        self.collapse_spaces = collapse_spaces

    def strip(self, html_text: str) -> str:
        # Remove script and style blocks (with their content)
        text = re.sub(r'<script\b[^>]*>.*?</script>', '', html_text, flags=re.DOTALL | re.IGNORECASE)
        text = re.sub(r'<style\b[^>]*>.*?</style>', '', text, flags=re.DOTALL | re.IGNORECASE)

        # Optionally remove comments
        if self.strip_comments:
            text = re.sub(r'<!--.*?-->', '', text, flags=re.DOTALL)

        # Convert <br> to newline
        if self.convert_br:
            text = re.sub(r'<br\s*/?>', '\n', text, flags=re.IGNORECASE)

        # Remove all remaining tags
        text = re.sub(r'<[^>]+>', '', text)

        # Decode HTML entities
        if self.decode_entities:
            text = html.unescape(text)

        # Collapse whitespace (optional)
        if self.collapse_spaces:
            text = re.sub(r'[ \t]+', ' ', text)
            text = re.sub(r'\n\s*\n', '\n\n', text)
            text = text.strip()

        return text

    def strip_file(self, filepath: str) -> str:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
        return self.strip(content)

def main():
    stripper = HTMLStripper()
    print("=== HTML Stripper ===")
    while True:
        print("\n1. Strip HTML from input text")
        print("2. Strip HTML from file")
        print("3. Toggle options (comments: {}, convert_br: {}, decode_entities: {}, collapse_spaces: {})".format(
            stripper.strip_comments, stripper.convert_br, stripper.decode_entities, stripper.collapse_spaces))
        print("4. Exit")
        choice = input("Choose: ").strip()
        if choice == '1':
            print("Enter HTML text (end with empty line):")
            lines = []
            while True:
                line = input()
                if line == '':
                    break
                lines.append(line)
            html_text = '\n'.join(lines)
            result = stripper.strip(html_text)
            print("\nProcessed result:\n", result)
        elif choice == '2':
            fname = input("Enter file path: ").strip()
            try:
                result = stripper.strip_file(fname)
                print("\nProcessed result:\n", result)
            except FileNotFoundError:
                print("File not found.")
            except Exception as e:
                print(f"Error: {e}")
        elif choice == '3':
            # Toggle options
            stripper.strip_comments = not stripper.strip_comments
            stripper.convert_br = not stripper.convert_br
            stripper.decode_entities = not stripper.decode_entities
            stripper.collapse_spaces = not stripper.collapse_spaces
            print("Options toggled.")
        elif choice == '4':
            print("Goodbye!")
            break
        else:
            print("Invalid choice.")

if __name__ == "__main__":
    main()

File: HtmlStripper.java
// HtmlStripper.java
import java.io.*;
import java.util.*;
import java.util.regex.*;

public class HtmlStripper {
    private boolean stripComments = true;
    private boolean convertBr = true;
    private boolean decodeEntities = true;
    private boolean collapseSpaces = true;

    public String strip(String htmlText) {
        String text = htmlText;
        text = text.replaceAll("(?is)<script\\b[^>]*>.*?</script>", "");
        text = text.replaceAll("(?is)<style\\b[^>]*>.*?</style>", "");
        if (stripComments)
            text = text.replaceAll("<!--.*?-->", "");
        if (convertBr)
            text = text.replaceAll("(?i)<br\\s*/?>", "\n");
        text = text.replaceAll("<[^>]+>", "");
        if (decodeEntities)
            text = decodeEntities(text);
        if (collapseSpaces) {
            text = text.replaceAll("[ \\t]+", " ");
            text = text.replaceAll("\\n\\s*\\n", "\n\n");
            text = text.trim();
        }
        return text;
    }

    private String decodeEntities(String s) {
        Map<String, String> entities = new HashMap<>();
        entities.put("&nbsp;", " "); entities.put("&amp;", "&");
        entities.put("&lt;", "<"); entities.put("&gt;", ">");
        entities.put("&quot;", "\""); entities.put("&#39;", "'");
        entities.put("&apos;", "'"); entities.put("&copy;", "©");
        entities.put("&reg;", "®"); entities.put("&trade;", "™");
        entities.put("&euro;", "€"); entities.put("&pound;", "£");
        entities.put("&yen;", "¥"); entities.put("&cent;", "¢");
        entities.put("&sect;", "§"); entities.put("&deg;", "°");
        for (Map.Entry<String, String> e : entities.entrySet())
            s = s.replace(e.getKey(), e.getValue());
        return s;
    }

    public String stripFile(String filepath) throws IOException {
        StringBuilder sb = new StringBuilder();
        try (BufferedReader br = new BufferedReader(new FileReader(filepath))) {
            String line;
            while ((line = br.readLine()) != null)
                sb.append(line).append("\n");
        }
        return strip(sb.toString());
    }

    public static void main(String[] args) throws IOException {
        HtmlStripper stripper = new HtmlStripper();
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        System.out.println("=== HTML Stripper ===");
        while (true) {
            System.out.println("\n1. Strip HTML from input text");
            System.out.println("2. Strip HTML from file");
            System.out.printf("3. Toggle options (comments: %b, convert_br: %b, decode_entities: %b, collapse_spaces: %b)\n",
                    stripper.stripComments, stripper.convertBr, stripper.decodeEntities, stripper.collapseSpaces);
            System.out.println("4. Exit");
            System.out.print("Choose: ");
            String choice = reader.readLine().trim();
            switch (choice) {
                case "1":
                    System.out.println("Enter HTML text (end with empty line):");
                    StringBuilder sb = new StringBuilder();
                    while (true) {
                        String line = reader.readLine();
                        if (line.isEmpty()) break;
                        sb.append(line).append("\n");
                    }
                    String htmlText = sb.toString();
                    String result = stripper.strip(htmlText);
                    System.out.println("\nProcessed result:\n" + result);
                    break;
                case "2":
                    System.out.print("Enter file path: ");
                    String fname = reader.readLine().trim();
                    try {
                        result = stripper.stripFile(fname);
                        System.out.println("\nProcessed result:\n" + result);
                    } catch (FileNotFoundException e) {
                        System.out.println("File not found.");
                    } catch (Exception e) {
                        System.out.println("Error: " + e.getMessage());
                    }
                    break;
                case "3":
                    stripper.stripComments = !stripper.stripComments;
                    stripper.convertBr = !stripper.convertBr;
                    stripper.decodeEntities = !stripper.decodeEntities;
                    stripper.collapseSpaces = !stripper.collapseSpaces;
                    System.out.println("Options toggled.");
                    break;
                case "4":
                    System.out.println("Goodbye!");
                    return;
                default:
                    System.out.println("Invalid choice.");
            }
        }
    }
}
